src/IR/SubstMagic.hs

```diff
@@ -0,0 +1,44 @@
+-- | Substitute AST nodes with magical primitives.
+{-# LANGUAGE OverloadedStrings #-}
+module IR.SubstMagic
+  ( substMagic
+  ) where
+
+import qualified IR.IR                         as I
+
+import           Data.Data                      ( Proxy(..) )
+import           Data.Generics                  ( Data(..)
+                                                , everywhere
+                                                , mkT
+                                                )
+
+{- | Substitute AST nodes with magical primitives.
+
+Implemented as a syb-style generic tree traversal.
+
+Example usage:
+
+@@
+-- Given:
+myExpr :: Expr Poly.Type
+
+-- Do:
+substMagic (Proxy :: Proxy Poly.Type) myExpr
+
+-- Given:
+myProm :: Program Annotated.Type
+
+-- Do:
+substMagic (Proxy :: Proxy Annotated.Type) myProg
+@@
+-}
+substMagic :: (Data t, Data a) => Proxy t -> a -> a
+substMagic p = everywhere $ mkT $ substMagicExpr p
```

Comment:
Is this pass going to be inserted before lowering AST?

Reply:
This pass as it's currently written would perform the substitution on the already-lowered IR.

I'm considering changing strategy and doing this as part of the lowering phase instead.